### PR TITLE
Add AI/LLM generated tests

### PR DESCRIPTION
**Pull request by nstadigs**
I gave convention.md together with the already created tests to GPT-4o.

Didn't get to datetime, duration and JSON because my free tier ran out for the day.

These needs to be validated!

### CODE REVIEW

File: homie5/values/string.yml [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  #   input_data: "a" * 268435456  # Represents a string at max length
+  #   output_data: "a" * 268435456
+  #   valid: true
+
+  # - description: Invalid string exceeding maximum length
+  #   testtype: propertyvaluestring
+  #   definition
+  #     datatype: string
+  #   input_data: "a" * 268435457  # Exceeds max length by one character
+  #   valid: false
```
Comment (nstadigs) [inline homie5/values/string.yml]: Can these tests be represented in yaml somehow?

**Comment by nstadigs** [comment]
rofl @Tieske already did this https://github.com/homieiot/homie-testsuite/pull/3 :laughing: 

Closing

**Comment by Tieske** [comment]
some of mine were placeholder files, so maybe we can still use some of this as well

**Comment by nstadigs** [comment]
Alright. When yours are merged I can rebase these on top

**Comment by Tieske** [comment]
Can you review the other PR? if ok, I'll merge it

**Comment by Tieske** [comment]
can you rebase this? the formats are duplicates by now, but the values are probably filling in the placeholders I put in there

**Comment by nstadigs** [comment]
> can you rebase this? the formats are duplicates by now, but the values are probably filling in the placeholders I put in there

Yeah, I plan to continue on this when I get some spare time. It's in draft mode for now

File: homie5/formats/float.yml [hunk 2]
```diff
@@ -1,5 +1,128 @@
...
+      datatype: float
+      format: "0:10:20"
+    valid: false
+
+  - description: missing range boundaries are rejected
+    testtype: propertydescription
+    definition:
+      datatype: float
+      format: ":"
+    valid: false
```
Comment (Tieske) [inline homie5/formats/float.yml]:
```suggestion
  - description: missing range boundaries are accepted
    testtype: propertydescription
    definition:
      datatype: float
      format: ":"
    valid: true
```

This one should be accepted, since it is the default format.

File: homie5/formats/float.yml [hunk 3]
```diff
@@ -1,5 +1,128 @@
...
+      format: "10.1:50.9"
+    valid: true
+
+  - description: proper format with step value is accepted
+    testtype: propertydescription
+    definition:
+      datatype: float
+      format: "0:100:0.5"
+    valid: true
+
```
Comment (Tieske) [inline homie5/formats/float.yml:31]:
```suggestion

  - description: proper format with only step value is accepted
    testtype: propertydescription
    definition:
      datatype: float
      format: "::0.5"
    valid: true

```

File: homie5/formats/float.yml
Comment (Tieske) [inline homie5/formats/float.yml:1]: What's missing is the full float format (exp notation, "." as decimal separator, etc.)

File: homie5/formats/integer.yml [hunk 4]
```diff
@@ -1,5 +1,107 @@
...
+      datatype: integer
+      format: "0:10:20"
+    valid: false
+
+  - description: missing range boundaries is rejected
+    testtype: propertydescription
+    definition:
+      datatype: integer
+      format: ":"
+    valid: false
```
Comment (Tieske) [inline homie5/formats/integer.yml]:
```suggestion
  - description: missing range boundaries is accepted
    testtype: propertydescription
    definition:
      datatype: integer
      format: ":"
    valid: true
```

File: homie5/formats/float.yml [hunk 5]
```diff
@@ -1,5 +1,128 @@
...
+      format: "0:100:0"
+    valid: false
+
+  - description: step value larger than range is rejected
+    testtype: propertydescription
+    definition:
+      datatype: float
+      format: "0:10:20"
+    valid: false
+
```
Comment (Tieske) [inline homie5/formats/float.yml]:
```suggestion
  - description: step value larger than range is accepted
    testtype: propertydescription
    definition:
      datatype: float
      format: "0:10:20"
    valid: true

```
The only valid value is "0" in this case. If not valid then it would collide with the `single-point range with step value is accepted` test above.

File: homie5/formats/integer.yml [hunk 6]
```diff
@@ -1,5 +1,107 @@
...
+      format: "10:50"
+    valid: true
+
+  - description: proper format with step value is accepted
+    testtype: propertydescription
+    definition:
+      datatype: integer
+      format: "0:100:5"
+    valid: true
+
```
Comment (Tieske) [inline homie5/formats/integer.yml:31]:
```suggestion

  - description: proper format with only step value is accepted
    testtype: propertydescription
    definition:
      datatype: integer
      format: "::5"
    valid: true

```

File: homie5/formats/integer.yml [hunk 7]
```diff
@@ -1,5 +1,107 @@
...
+      datatype: integer
+      format: "0:100:0"
+    valid: false
+
+  - description: step value larger than range is rejected
+    testtype: propertydescription
+    definition:
+      datatype: integer
+      format: "0:10:20"
+    valid: false
```
Comment (Tieske) [inline homie5/formats/integer.yml]:
```suggestion
  - description: step value larger than range is accepted
    testtype: propertydescription
    definition:
      datatype: integer
      format: "0:10:20"
    valid: true
```
collides with `single point range with step value is accepted` above

File: homie5/values/enum.yml [hunk 8]
```diff
@@ -1,3 +1,120 @@
...
+    valid: true
+
+  - description: invalid enum value when format is empty
+    testtype: propertyvalueenum
+    definition:
+      datatype: enum
+      format: ""
+    input_data: "anyvalue"
+    valid: false
+
```
Comment (Tieske) [inline homie5/values/enum.yml]:
```suggestion
```
the format is invalid, cannot test this

File: homie5/values/enum.yml [hunk 9]
```diff
@@ -1,3 +1,120 @@
...
+  - description: valid enum with special characters
+    testtype: propertyvalueenum
+    definition:
+      datatype: enum
+      format: "on/off,standby,active"
+    input_data: "on/off"
+    output_data: "on/off"
+    valid: true
+
```
Comment (Tieske) [inline homie5/values/enum.yml]:
```suggestion
```

these are invalid formats

File: homie5/formats/integer.yml [hunk 10]
```diff
@@ -1,5 +1,107 @@
...
+      format: "-100:0:10"
+    valid: true
+
+  - description: single point range with step value is accepted
+    testtype: propertydescription
+    definition:
+      datatype: integer
+      format: "5:5:1"
+    valid: true
+
```
Comment (Tieske) [inline homie5/formats/integer.yml:94]:
```suggestion

  - description: exponent notation in minimum is rejected
    testtype: propertydescription
    definition:
      datatype: integer
      format: "5e1:5000:10"
    valid: false

  - description: exponent notation in maximum is rejected
    testtype: propertydescription
    definition:
      datatype: integer
      format: "50:5e5:10"
    valid: false

  - description: exponent notation in step is rejected
    testtype: propertydescription
    definition:
      datatype: integer
      format: "50:5000:1e2"
    valid: false

  - description: only negative sign in minimum is rejected
    testtype: propertydescription
    definition:
      datatype: integer
      format: "-:10:1"
    valid: false
  
  - description: only negative sign in maximum is rejected
    testtype: propertydescription
    definition:
      datatype: integer
      format: "-5:-:1"
    valid: false

  - description: only negative sign in step is rejected
    testtype: propertydescription
    definition:
      datatype: integer
      format: "5:10:-"
    valid: false

```

File: homie5/values/float.yml
Comment (Tieske) [inline homie5/values/float.yml:1]: to add: exponent notation and absence of leading zero; eg. "-.15" should be accepted

File: homie5/values/string.yml [hunk 11]
```diff
@@ -1,3 +1,66 @@
...
+    valid: true
+
+  - description: Valid string with special characters
+    testtype: propertyvaluestring
+    definition:
+      datatype: string
+    input_data: "!@#$%^&*()_+|"
+    output_data: "!@#$%^&*()_+|"
+    valid: true
+
```
Comment (Tieske) [inline homie5/values/string.yml:19]:
```suggestion

  - description: Valid empty string value
    testtype: propertyvaluestring
    definition:
      datatype: string
    input_data: "\x00"
    output_data: ""
    valid: true

```

**Review by Tieske — CHANGES_REQUESTED** [review]
Here's a start for a review. Though incomplete.

floats need tests on exponential notation and missing lead-zero's, eg ".001" and "-.1" are valid (imho)

also need to verify the range tests still.

**Comment by nstadigs** [comment]
Don't see these test files as complete for the included `datatype`s. I think missing tests can be added later. But we shouldn't add any tests that are wrong.

**Comment by Tieske** [comment]
Agreed. If you can review and commit my suggestions, then I think we can merge this set.

File: homie5/values/enum.yml (hunk 8 above)
Comment (nstadigs) [reply homie5/values/enum.yml]: nice catch

**Comment by Tieske** [comment]
please do not force-push/rebase just add new commits such that it is easier to track which comments have been resolved. That prevent reviewers from having to review the whole thing from scratch again.

We can still squash just before merging.

Can you resolve the conversations/comments that you updated? then I'll give it another review.

**Comment by nstadigs** [comment]
> please do not force-push/rebase just add new commits such that it is easier to track which comments have been resolved. That prevent reviewers from having to review the whole thing from scratch again.
> 
> We can still squash just before merging.
> 
> Can you resolve the conversations/comments that you updated? then I'll give it another review.

Oops, sorry. I'm used to Gitlab at work which handles this perfectly, so I didn't think about it

**Comment by nstadigs** [comment]
I know I force pushed again. I made a reset to the revision that you commented on to get that back into history. I then made another commit with the changes on top of it, to see if I could fix the comment references here.

**Review by Tieske — APPROVED** [review]
lgtm
